### PR TITLE
lower bounded target q

### PR DESCRIPTION
**Pull request by le-horizon**
minimum change for lower bounded value target (for episodic return, goal distance return, and n-step bootstrapped return)

### CODE REVIEW

File: alf/utils/value_ops.py [hunk 1]
```diff
@@ -118,6 +118,72 @@ def action_importance_ratio(action_distribution,
     return importance_ratio, importance_ratio_clipped
 
 
+def generalized_advantage_estimation(rewards,
```
Comment (le-horizon) [inline alf/utils/value_ops.py]: I didn't change ``generalized_advantage_estimation``, only moved it closer to action_importance_ratio, git messed up the diff..

File: alf/utils/value_ops.py [hunk 2]
```diff
@@ -180,101 +246,97 @@ def discounted_return(rewards, values, step_types, discounts, time_major=True):
     return rets.detach()
 
 
-def one_step_discounted_return(rewards, values, step_types, discounts):
-    """Calculate the one step discounted return  for the first T-1 steps.
+def first_step_future_discounted_returns(rewards,
```
Comment (le-horizon) [inline alf/utils/value_ops.py:249]: This ``first_step_future_discounted_returns`` function is the only new thing in this file.

File: alf/utils/value_ops.py [hunk 3]
```diff
@@ -180,101 +246,97 @@ def discounted_return(rewards, values, step_types, discounts, time_major=True):
...
-    "High-Dimensional Continuous Control Using Generalized Advantage Estimation"
-    by John Schulman, Philipp Moritz et al.
-    See https://arxiv.org/abs/1506.02438 for full paper.
+    return rets.detach()
 
-    The difference between this function and the one tf_agents.utils.value_ops
-    is that the accumulated_td is reset to 0 for is_last steps in this function.
 
-    Define abbreviations:
+def one_step_discounted_return(rewards, values, step_types, discounts):
```
Comment (le-horizon) [inline alf/utils/value_ops.py:312]: No changes were made to this ``one_step_discounted_return`` function.

File: alf/algorithms/td_loss.py [hunk 4]
```diff
@@ -230,6 +237,221 @@ def _summarize(v, r, td, suffix):
...
+            gamma .. use_retrace: pass through to TDLoss.
+            lb_target_q: between 0 and 1.  When not zero, use this mixing rate for the
+                lower bounded value target.  Only supports batch_length == 2, one step td.
+            default_return: Keep it the same as replay_buffer.default_return to plot to
+                tensorboard episodic_discounted_return only for the timesteps whose
+                episode already ended.
+            improve_w_goal_return: Use return calculated from the distance to hindsight
+                goals.  Only supports batch_length == 2, one step td.
+            improve_w_nstep_bootstrap: Look ahead 2 to n steps, and take the largest
+                bootstrapped return to lower bound the value target of the 1st step.
```
Comment (le-horizon) [inline alf/algorithms/td_loss.py:286]: add formula

File: alf/algorithms/ddpg_algorithm.py [hunk 5]
```diff
@@ -40,9 +40,20 @@
...
+        "discount",
+        "action",
+        "action_distribution",
+        "actor_loss",
+        "critic",
+        # Optional fields for value target lower bounding or Hindsight relabeling.
+        # TODO: Extract these into a HerAlgorithm wrapper for easier adoption of HER.
+        "discounted_return",
+        "future_distance",
+        "her"
```
Comment (le-horizon) [inline alf/algorithms/ddpg_algorithm.py:56]: "is_her"

File: alf/algorithms/td_loss.py (hunk 4 above)
Comment (le-horizon) [reply alf/algorithms/td_loss.py:286]: Done.

**Review by le-horizon — COMMENTED** [review]
@emailweixu and @hnyu , I've separated out the HER related logic (the part that move batch_info fields into alg_info and loss calculation) into a new file her_algorithms.

Let me know how you like this version.

Thanks,
Le
